### PR TITLE
destination-redshift: Heartbeats for bastion and other minor bugfixes

### LINKED ISSUE #33643: Redshift Dv2 Timeouts (`marked as broken because of SQLSTATE`)
Consider these connections ([1](https://cloud.airbyte.com/workspaces/502c1fb8-d9ce-4763-99ec-578b3af34ceb/connections/c86ba404-e1c1-463e-9cb1-603ba7a68e04/job-history#6796783::1
), [2](https://cloud.airbyte.com/workspaces/502c1fb8-d9ce-4763-99ec-578b3af34ceb/connections/c86ba404-e1c1-463e-9cb1-603ba7a68e04/job-history#6766286::0))

Slack ([1](https://airbytehq-team.slack.com/archives/C0690DFUDGS/p1702915718531749?thread_ts=1702299183.996059&cid=C0690DFUDGS), [2](https://airbytehq-team.slack.com/archives/C0690DFUDGS/p1703001199431139))

> To me it looks like the connection keep alive or similar is not working properly.
Here are two initial sync attempts that ran into a connection timeout while filling the destination tables.

```
2023-12-18 11:17:01 destination > COMMIT;
2023-12-18 11:27:00 destination > 2023-12-18 11:27:00 WARN o.a.s.c.u.l.LoggingUtils(warn):618 - exceptionCaught(ClientSessionImpl[airbyte@xxx.eu-central-1.compute.amazonaws.com/xxx:22])[state=Opened] SocketException: Connection reset
2023-12-18 11:27:00 destination > 2023-12-18 11:27:00 WARN c.z.h.p.ProxyConnection(checkException):177 - HikariPool-1 - Connection com.amazon.redshift.jdbc.RedshiftConnectionImpl@2a7e15eb marked as broken because of SQLSTATE(08006), ErrorCode(0)
2023-12-18 11:27:00 destination > com.amazon.redshift.util.RedshiftException: An I/O error occurred while sending to the backend.
2023-12-18 11:27:00 destination >         at com.amazon.redshift.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:402) ~[redshift-jdbc42-2.1.0.23.jar:?]
2023-12-18 11:27:00 destination >         at com.amazon.redshift.jdbc.RedshiftStatementImpl.executeInternal(RedshiftStatementImpl.java:515) ~[redshift-jdbc42-2.1.0.23.jar:?]
2023-12-18 11:27:00 destination >         at com.amazon.redshift.jdbc.RedshiftStatementImpl.execute(RedshiftStatementImpl.java:436) ~[redshift-jdbc42-2.1.0.23.jar:?]
2023-12-18 11:27:00 destination >         at com.amazon.redshift.jdbc.RedshiftStatementImpl.executeWithFlags(RedshiftStatementImpl.java:377) ~[redshift-jdbc42-2.1.0.23.jar:?]
2023-12-18 11:27:00 destination >         at com.amazon.redshift.jdbc.RedshiftStatementImpl.executeCachedSql(RedshiftStatementImpl.java:363) ~[redshift-jdbc42-2.1.0.23.jar:?]
2023-12-18 11:27:00 destination >         at com.amazon.redshift.jdbc.RedshiftStatementImpl.executeWithFlags(RedshiftStatementImpl.java:340) ~[redshift-jdbc42-2.1.0.23.jar:?]
2023-12-18 11:27:00 destination >         at com.amazon.redshift.jdbc.RedshiftStatementImpl.execute(RedshiftStatementImpl.java:330) ~[redshift-jdbc42-2.1.0.23.jar:?]
2023-12-18 11:27:00 destination >         at com.zaxxer.hikari.pool.ProxyStatement.execute(ProxyStatement.java:94) ~[HikariCP-5.0.1.jar:?]
2023-12-18 11:27:00 destination >         at com.zaxxer.hikari.pool.HikariProxyStatement.execute(HikariProxyStatement.java) ~[HikariCP-5.0.1.jar:?]
2023-12-18 11:27:00 destination >         at io.airbyte.cdk.db.jdbc.JdbcDatabase.lambda$execute$0(JdbcDatabase.java:49) ~[airbyte-cdk-core-0.7.3.jar:?]
2023-12-18 11:27:00 destination >         at io.airbyte.cdk.db.jdbc.DefaultJdbcDatabase.execute(DefaultJdbcDatabase.java:46) ~[airbyte-cdk-core-0.7.3.jar:?]
2023-12-18 11:27:00 destination >         at io.airbyte.cdk.db.jdbc.JdbcDatabase.execute(JdbcDatabase.java:49) ~[airbyte-cdk-core-0.7.3.jar:?]
2023-12-18 11:27:00 destination >         at io.airbyte.cdk.integrations.destination.jdbc.typing_deduping.JdbcDestinationHandler.execute(JdbcDestinationHandler.java:76) ~[airbyte-cdk-db-destinations-0.7.3.jar:?]
2023-12-18 11:27:00 destination >         at io.airbyte.integrations.base.destination.typing_deduping.TypeAndDedupeTransaction.executeTypeAndDedupe(TypeAndDedupeTransaction.java:39) ~[airbyte-cdk-typing-deduping-0.7.3.jar:?]
2023-12-18 11:27:00 destination >         at io.airbyte.integrations.base.destination.typing_deduping.DefaultTyperDeduper.lambda$typeAndDedupeTask$2(DefaultTyperDeduper.java:220) ~[airbyte-cdk-typing-deduping-0.7.3.jar:?]
2023-12-18 11:27:00 destination >         at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1768) [?:?]
2023-12-18 11:27:00 destination >         at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136) [?:?]
2023-12-18 11:27:00 destination >         at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635) [?:?]
2023-12-18 11:27:00 destination >         at java.base/java.lang.Thread.run(Thread.java:833) [?:?]
2023-12-18 11:27:00 destination > Caused by: java.io.EOFException: The server closed the connection.
```

We likely have a keep-alive/timeout issue with the Redshift driver/client giving up when there are Very Slow (tm) queries.

Highly relevant - https://docs.aws.amazon.com/redshift/latest/mgmt/jdbc20-install.html#configure-tcp-keepalives-jdbc20 (default), but we probably need https://stackoverflow.com/questions/54804275/hikaripool-1-connection-marked-as-broken-because-of-sqlstate08s01-errorcode

### LINKED ISSUE #33900: Source S3/S3 File <> Redshift V2: final tables are `NULL`ed out (Case Sensitivity issues with SUEPR)
See https://airbytehq-team.slack.com/archives/C0690DFUDGS/p1703278044226589 for more screenshots + logs

From S3 source or file source with S3 file, Redshift V2 destination final tables only have `NULL` values

Scoping down the problem:
- Equivalent V1 sync works fine
- MySQL source <> Redshift V2 destination works fine
- `_airbyte_data` in V2 raw tables looks OK
- Observed with loading via S3 staging, no data on standard inserts

### LINKED ISSUE #33945: Redshift NullPointerException in hasUnprocessedRecords
A Bug from our beta test users ([slack](https://airbytehq-team.slack.com/archives/C0690DFUDGS/p1704388357215549?thread_ts=1704317572.946759&cid=C0690DFUDGS))

```
2024-01-04 13:16:59 destination > 	Caused by: java.lang.NullPointerException: Cannot invoke "io.airbyte.integrations.base.destination.typing_deduping.DestinationHandler$InitialRawTableState.hasUnprocessedRecords()" because the return value of "java.util.Map.get(Object)" is null
2024-01-04 13:16:59 destination > 		at io.airbyte.integrations.base.destination.typing_deduping.DefaultTyperDeduper.lambda$typeAndDedupe$6(DefaultTyperDeduper.java:273) ~[airbyte-cdk-typing-deduping-0.10.2.jar:?]
2024-01-04 13:16:59 destination > 		at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:178) ~[?:?]
2024-01-04 13:16:59 destination > 		at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1625) ~[?:?]
2024-01-04 13:17:00 destination > 		at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509) ~[?:?]
2024-01-04 13:17:00 destination > 		at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499) ~[?:?]
2024-01-04 13:17:00 destination > 		at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:150) ~[?:?]
2024-01-04 13:17:00 destination > 		at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:173) ~[?:?]
2024-01-04 13:17:00 destination > 		at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234) ~[?:?]
2024-01-04 13:17:00 destination > 		at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:596) ~[?:?]
2024-01-04 13:17:00 destination > 		at io.airbyte.integrations.base.destination.typing_deduping.DefaultTyperDeduper.typeAndDedupe(DefaultTyperDeduper.java:285) ~[airbyte-cdk-typing-deduping-0.10.2.jar:?]
2024-01-04 13:17:00 destination > 		at io.airbyte.cdk.integrations.destination.staging.GeneralStagingFunctions.lambda$onCloseFunction$1(GeneralStagingFunctions.java:123) ~[airbyte-cdk-db-destinations-0.10.2.jar:?]
2024-01-04 13:17:00 destination > 		at io.airbyte.cdk.integrations.destination.staging.StagingConsumerFactory.lambda$createAsync$0(StagingConsumerFactory.java:118) ~[airbyte-cdk-s3-destinations-0.10.2.jar:?]
```

Full sync log here: 
[f05d2773_e10a_48b4_82c2_8b3f88d23f98_job_8612_attempt_13_txt.txt](https://github.com/airbytehq/airbyte/files/13833578/f05d2773_e10a_48b4_82c2_8b3f88d23f98_job_8612_attempt_13_txt.txt)

### PR DESCRIPTION
**Pull request by gisripa**
## What
Fixes: #33945 
Fixes: #33900

Potentially Fixes: #33643

## How
See CDK changes #33727
Adding addition redshift connectTimeout from default 10s

### CODE REVIEW

**Review by edgao — APPROVED** [review]
maybe worth also releasing bigquery+snowflake for the T+D setup fix?

**Comment by gisripa** [comment]
> maybe worth also releasing bigquery+snowflake for the T+D setup fix?

Oh right. DefaultTD is a project dependency right ? so just version bumps should do ?

**Comment by edgao** [comment]
> just version bumps should do

yep

File: airbyte-integrations/connectors/destination-redshift/src/test-integration/resources/sqlgenerator/alltypes_expectedrecords_final.jsonl [hunk 1]
```diff
@@ -5,3 +5,4 @@
 // Note that for numbers where we parse the value to JSON (struct, array, unknown) we lose precision.
 // But for numbers where we create a NUMBER column, we do not lose precision (see the `number` column).
 {"id1": 5, "id2": 100, "updated_at": "2023-01-01T01:00:00.000000Z", "number": 67.174118, "struct": {"nested_number": 67.174118}, "array": [67.174118], "unknown": 67.174118, "_airbyte_extracted_at": "2023-01-01T00:00:00.000000Z", "_airbyte_meta": {"errors": []}}
+{"id1": 6, "id2": 100, "updated_at": "2023-01-01T01:00:00.000000Z", "iamacasesensitivecolumnname": "Case senstive value", "_airbyte_extracted_at": "2023-01-01T00:00:00.000000Z", "_airbyte_meta": {"errors": []}}
```
Comment (edgao) [inline airbyte-integrations/connectors/destination-redshift/src/test-integration/resources/sqlgenerator/alltypes_expectedrecords_final.jsonl:8]:
```suggestion
// Note that redshift downcases IAmACaseSensitiveColumnName to all lowercase
{"id1": 6, "id2": 100, "updated_at": "2023-01-01T01:00:00.000000Z", "iamacasesensitivecolumnname": "Case senstive value", "_airbyte_extracted_at": "2023-01-01T00:00:00.000000Z", "_airbyte_meta": {"errors": []}}
```

File: airbyte-integrations/connectors/destination-redshift/src/main/java/io/airbyte/integrations/destination/redshift/RedshiftStagingS3Destination.java [hunk 2]
```diff
@@ -143,7 +144,24 @@ protected NamingConventionTransformer getNamingResolver() {
...
+    // connectTimeout is different from Hikari pool's connectionTimout, driver defaults to 10seconds so
+    // increase it to match hikari's default
+    connectionOptions.put("connectTimeout", "120");
+    // HikariPool properties
+    // https://github.com/brettwooldridge/HikariCP?tab=readme-ov-file#frequently-used
+    // connectionTimeout is set explicitly to 2 minutes when creating data source.
+    // Do aggressive keepAlive with minimum allowed value, this only applies to connection sitting idle
+    // in the pool.
+    connectionOptions.put("keepaliveTime", Long.toString(Duration.ofSeconds(30).toMillis()));
+    connectionOptions.put("connectionInitSql", "SET enable_case_sensitive_identifier to TRUE;");
```
Comment (edgao) [inline airbyte-integrations/connectors/destination-redshift/src/main/java/io/airbyte/integrations/destination/redshift/RedshiftStagingS3Destination.java:162]:
```suggestion
```

**Review by edgao — APPROVED** [review]
two nits, otherwise lgtm!

**Pull request merged**